### PR TITLE
String as define issue

### LINKED ISSUE #4153: The Python runner is not properly escaping string parameters when used with Verilator
Cocotb 1.9.1, Ubuntu 24.04 64 but, Verilator 5.024, System python 3.12.3. 

I have SystemVerilog design with string parameter and I'm using python runner like this: 

```
    param = {
        'MEM_MACRO_NAME': 'example_mem_cell',
        'WIDTH': 16,
    }

    runner.build(
       <...>
        parameters=param,
        <...>
    )

   runner.test(<...>)
```

This launches verilator with argument `-GMEM_MACRO_NAME=example_mem_cell`, and results in error:

```
%Error: PATH_TO_FILE_REDACTED.sv: Illegal character in decimal constant: e
    8 |   parameter MEM_MACRO_NAME = "example_mem_cell",  
    
```

This is explained in [verilator documentation](https://verilator.org/guide/latest/exe_verilator.html#cmdoption-G-name):

> -G\<name\>=\<value\>
> Overwrites the given parameter of the top-level module. The value is limited to basic data literals:
> <...>
> 
> Strings
> Strings must be in double quotes (“”). 
> They must be escaped properly on the command line, e.g., as -GSTR="\\"My String\\"" or -GSTR='"My String"'.

So it seems like python runner is not passing string parameters correctly. 

I managed to get it working with     
`
        'MEM_MACRO_NAME': '\"example_mem_cell\"',
`
but python runner probably should do it automatically in this case.

### PR DESCRIPTION
**Pull request by vborchsh**
Closes #4153

String as define doesn't work. At least, for Questa and Icarus. I tried some tricks, but no luck. Is there a way to do it?

Thanks.

### CODE REVIEW

**Comment by ktbarrett** [comment]
All failing tests are due to syntax error.

**Comment by vborchsh** [comment]
Ah, sorry, I was not clear. And I have missed commas :)

> All failing tests are due to syntax error.

This is true, because of wrong formatting of string define parameter for the simulators. E.g:

```
iverilog ...blabla... -DDEFINE_PATH=path/to/some/file.wow
```

This part must be like `-DDEFINE_PATH="path/to/some/file.wow"`

The same situation for Questa:

```
vlog ...blabla... +define+DEFINE_PATH=path/to/some/file.wow
```
to `+define+DEFINE_PATH="path/to/some/file.wow"`

I played a bit with combinations of lists, tuples, strings and so on. But no luck. Maybe there is a way to do it...

**Comment by ktbarrett** [comment]
I think the real answer is to create a function that takes the value and serializes it appropriately and use that instead of `as_tcl_value`.

**Review by themperek — APPROVED** [review]

**Comment by themperek** [comment]
Sorry I approve by mistake.

**Comment by themperek** [comment]
Would this work `"DEFINE_PATH": "\"path/to/some/file.wow\""` ?

**Comment by Lockedloop** [comment]
Removing the `as_tcl_value` completely works for me.
```
    @staticmethod
    def _get_define_options(defines: Mapping[str, object]) -> Command:
        return [
            f"+define+{name}={value}"
            for name, value in defines.items()
        ]
```
I could be missing something, but seeing that `_get_parameter_options` does the same thing in the current version, it seems to me as a quick fix. Does anyone know why was the `as_tcl_value` introduced in the first place?

**Comment by ktbarrett** [comment]
@Lockedloop Good to know. `as_tcl_value` was added to serialize values in expected TCL formatting since some simulators need that, but it's obviously not appropriate here. I'll try to get around to finishing this PR off if @davekeeshan doesn't in the next couple days.

**Comment by Lockedloop** [comment]
> @Lockedloop Good to know. `as_tcl_value` was added to serialize values in expected TCL formatting since some simulators need that, but it's obviously not appropriate here. I'll try to get around to finishing this PR off if @vborchsh doesn't in the next couple days.

Great, thanks a million!

**Comment by vborchsh** [comment]
> @Lockedloop Good to know. `as_tcl_value` was added to serialize values in expected TCL formatting since some simulators need that, but it's obviously not appropriate here. I'll try to get around to finishing this PR off if @vborchsh doesn't in the next couple days.

Sorry, just found this thread. Completely fine for me if you finish this :)

**Comment by ktbarrett** [comment]
Not sure what the Riviera and Xcelium issues are, but the icarus ones are because the one on apt is too old, I'll have to split that into a separate test and put a version ifdef on it. But I ca. See it being done soon.

**Comment by cmarqu** [comment]
Xcelium appears to need more quoting. Maybe one that `as_tcl_value` has done so far...

**Comment by ktbarrett** [comment]
Using `as_tcl_value` didn't help. The issue appears to be that the slashes need escaping.

**Comment by ktbarrett** [comment]
\\\\\\\"But\\\\ of\\\\ course!\\\\ It\\\\ needs\\\\ to\\\\ be\\\\ `+define+DEFINE_PATH=\\\"path/to/some\\\\\\\"(random\\ quote)/file.wow\\\"`\\\\\\\"

**Comment by ktbarrett** [comment]
So Xcelium doesn't seem to support spaces in defines passed on the command line no matter how many backslahes I add. This must be a limitation of their option parser. And it doesn't seem to like `$error()`? I'll just leave the problem child be.

**Comment by cmarqu** [comment]
> So Xcelium doesn't seem to support spaces in defines passed on the command line no matter how many backslahes I add. This must be a limitation of their option parser. And it doesn't seem to like `$error()`? I'll just leave the problem child be.

@imphil Do you think Cadence would help us here?

**Comment by imphil** [comment]
> @imphil Do you think Cadence would help us here?

Absolutely. Can you write the text you'd like me to submit as issue to Cadence? I'll then copy-paste it over.

**Comment by ktbarrett** [comment]
Let's see where this is now.

**Comment by ktbarrett** [comment]
I will have to split the test into two so we can run test_runner on xcelium without also running the define test as xcelium's option parser prevents us from passing strings with whitespace correctly (it will be set to expect fail).

File: tests/pytest/test_runner.py [hunk 1]
```diff
@@ -57,16 +60,14 @@ def test_runner(parameters, pre_cmd, clean_build):
...
     if hdl_toplevel_lang == "verilog":
-        sources = [os.path.join(tests_dir, "designs", "runner", "runner.v")]
+        sources = [os.path.join(tests_dir, "designs", "runner", "runner.sv")]
         gpi_interfaces = ["vpi"]
     else:
         sources = [os.path.join(tests_dir, "designs", "runner", "runner.vhdl")]
         gpi_interfaces = [vhdl_gpi_interfaces]
 
     runner = get_runner(sim)
     compile_args = []
```
Comment (ktbarrett) [inline tests/pytest/test_runner.py:70]:
```suggestion
    compile_args = [VHDL("-v93")] if sim == "xcelium" else []
```

File: tests/pytest/test_runner.py [hunk 2]
```diff
@@ -128,7 +129,7 @@ def test_missing_libpython(monkeypatch):
         WIDTH_IN="8",
         WIDTH_OUT="8",
     )
-    build_args = ["-v93"] if sim_tool == "xcelium" else []
+    build_args = []
```
Comment (ktbarrett) [inline tests/pytest/test_runner.py:132]:
```suggestion
    build_args = [VHDL("-v93")] if sim_tool == "xcelium" else []
```

**Review by marlonjames — APPROVED** [review]

**Comment by vborchsh** [comment]
Well, it's became definitely more complicated, that I thought... Thanks for fix it!

**Comment by ktbarrett** [comment]
@vborchsh Of course. Sorry it took so long. We finally have a good environment in place to test commercial sims arbitrarily which is what made this happen. Thanks @imphil!
